pkg/cmd/kind/load/docker-image/docker-image.go

```diff
@@ -121,38 +126,40 @@ func runE(logger log.Logger, flags *flagpole, args []string) error {
 
 	// pick only the nodes that don't have the image
 	selectedNodes := []nodes.Node{}
-	for _, node := range candidateNodes {
-		id, err := nodeutils.ImageID(node, imageName)
-		if err != nil || id != imageID {
-			selectedNodes = append(selectedNodes, node)
-			logger.V(0).Infof("Image: %q with ID %q not yet present on node %q, loading...", imageName, imageID, node.String())
+	fns := []func() error{}
+	for i, imageName := range imageNames {
+		imageID := imageIDs[i]
+		for _, node := range candidateNodes {
+			id, err := nodeutils.ImageID(node, imageName)
+			if err != nil || id != imageID {
+				selectedNodes = append(selectedNodes, node)
+				logger.V(0).Infof("Image: %q with ID %q not yet present on node %q, loading...", imageName, imageID, node.String())
+			}
+		}
+		if len(selectedNodes) == 0 {
+			return nil
 		}
-	}
-
-	if len(selectedNodes) == 0 {
-		return nil
-	}
 
-	// Save the image into a tar
-	dir, err := fs.TempDir("", "image-tar")
-	if err != nil {
-		return errors.Wrap(err, "failed to create tempdir")
-	}
-	defer os.RemoveAll(dir)
-	imageTarPath := filepath.Join(dir, "image.tar")
+		// Save the image into a tar
+		dir, err := fs.TempDir("", "image-tar")
+		if err != nil {
+			return errors.Wrap(err, "failed to create tempdir")
+		}
+		defer os.RemoveAll(dir)
+		imageTarPath := filepath.Join(dir, "image.tar")
 
-	err = save(imageName, imageTarPath)
-	if err != nil {
-		return err
-	}
+		err = save(imageName, imageTarPath)
+		if err != nil {
+			return err
+		}
 
-	// Load the image on the selected nodes
-	fns := []func() error{}
-	for _, selectedNode := range selectedNodes {
-		selectedNode := selectedNode // capture loop variable
-		fns = append(fns, func() error {
-			return loadImage(imageTarPath, selectedNode)
-		})
+		// Load the image on the selected nodes
+		for _, selectedNode := range selectedNodes {
+			selectedNode := selectedNode // capture loop variable
+			fns = append(fns, func() error {
+				return loadImage(imageTarPath, selectedNode)
+			})
+		}
```

Comment:
potentially, this code can be made to run in parallel via UntilErrorConcurrent:
https://github.com/kubernetes-sigs/kind/blob/a2282e90b85656f44eb4ee82a25dbd78eb7b1a6c/pkg/errors/concurrent.go#L25

Reply:
alright. So, just to confirm, should I add the snippet of code performing the `// Save the image into a tar`, a part of all the functions.

So, the new code would look like:

```go
		// Load the image on the selected nodes
		for _, selectedNode := range selectedNodes {
			selectedNode := selectedNode // capture loop variable
			fns = append(fns, func() error {
				// Save the image into a tar
				dir, err := fs.TempDir("", "image-tar")
				if err != nil {
					return errors.Wrap(err, "failed to create tempdir")
				}
				defer os.RemoveAll(dir)
				imageTarPath := filepath.Join(dir, "image.tar")

				err = save(imageName, imageTarPath)
				if err != nil {
					return err
				}
				return loadImage(imageTarPath, selectedNode)
			})
		}
```
And then, at the end `return errors.UntilErrorConcurrent(fns)` will run all the `fns` (involving saving and loading) in parallel

Is that what you mean?

Reply:
alright. I made the changes and brought the code performing the "saving of image into tar" just along the code where loading of images is happening. (Just above `return loadImage(imageTarPath, selectedNode)`)

Reply:
saving + loading in the same `fn` "should" work as long as the docker engine is fine with that.